File: Algorithms/340. Longest Substring with At Most K Distinct Characters - Method 2.py
# Given a string s and an integer k, return the length of the longest substring of s that contains at most k distinct characters.


class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        end = 0
        output = 0
        letters = {}

        for i, n in enumerate(s):
            letters[n] = letters.get(n, 0) + 1
            
            while len(letters) > k:
                letters[s[end]] -= 1
                if letters[s[end]] == 0:
                    del letters[s[end]]
                end += 1
            
            output = max(output, i - end + 1)
        
        return output

File: Algorithms/340. Longest Substring with At Most K Distinct Characters - Method 2.java
// Given a string s and an integer k, return the length of the longest substring of s that contains at most k distinct characters.


class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int end = 0;
        int output = 0;
        Map<Character, Integer> letters = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            Character temp = s.charAt(i);
            letters.put(temp, letters.get(temp)==null ? 1 : letters.get(temp)+1);
            while (letters.size() > k) {
                temp = s.charAt(end);
                if (letters.get(temp) == 1) {
                    letters.remove(temp);
                }
                else {
                    letters.put(temp, letters.get(temp)-1);
                }
                end++;
            }

            output = Math.max(output, i - end + 1);
        }

        return output;
    }
}
